pkg/agent/proxy/types/types.go

```diff
@@ -56,3 +56,34 @@ func NewEndpointInfo(baseInfo *k8sproxy.BaseEndpointInfo) k8sproxy.Endpoint {
 }
 
 type EndpointsMap map[k8sproxy.ServicePortName]map[string]k8sproxy.Endpoint
+
+func (m EndpointsMap) Equal(o EndpointsMap) bool {
+	if m == nil && o == nil {
+		return true
+	}
+	if m == nil || o == nil {
+		return false
+	}
+	if len(m) != len(o) {
+		return false
+	}
+	for svcPortName, endpoints := range m {
+		otherEndpoints, exists := o[svcPortName]
+		if !exists {
+			return false
+		}
+		if len(endpoints) != len(otherEndpoints) {
+			return false
+		}
```

Comment:
I noticed that you did some `nil` checks for the outer map but not for the inner map. Is that intentional?

Reply:
`m` or `o` may be `nil` as the method could be used for add and delete case. The inner map won't be `nil` because the we never set `nil` to a `svcPortName`.